orttraining/orttraining/python/training/orttrainer.py

```diff
@@ -843,3 +854,306 @@ def _update_onnx_model_initializers(self, state_tensors):
+                state_dict_model_state_keys = set(state_dict[model][precision_key])
+                unexpected_keys = list(state_dict_model_state_keys - training_session_model_state_keys)
+                if len(unexpected_keys):
+                    raise RuntimeError("Unexpected keys: {} in state_dict[model][{}]".format(unexpected_keys, precision_key))
+                missing_keys = list(training_session_model_state_keys - state_dict_model_state_keys)
+                if len(missing_keys):
+                    raise RuntimeError("Missing keys: {} in state_dict[model][{}]".format(missing_keys, precision_key))
+
+        def _check_optimizer_key_mismatch(training_session_state_dict, state_dict):
+            optimizer = 'optimizer'
+
+            # check for model state key mismatch for the optimizer state_dict
+            unexpected_optimizer_model_state_keys = list(set(state_dict[optimizer]) - \
+                set(training_session_state_dict[optimizer]))
+            if len(unexpected_optimizer_model_state_keys) > 0:
+                raise RuntimeError("Unexpected keys: {} in state_dict[optimizer]"\
+                    .format(unexpected_optimizer_model_state_keys))
+            missing_optimizer_model_state_keys = list(set(training_session_state_dict[optimizer]) - \
+                set(state_dict[optimizer]))
+            if len(missing_optimizer_model_state_keys) > 0:
+                raise RuntimeError("Missing keys: {} in state_dict[optimizer]"\
+                    .format(missing_optimizer_model_state_keys))
+
+            # check for optimizer state keys mismatch
+            for model_state_key in training_session_state_dict[optimizer]:
+                training_session_optimizer_state_keys = set(training_session_state_dict[optimizer][model_state_key])
+                state_dict_optimizer_state_keys = set(state_dict[optimizer][model_state_key])
+                unexpected_keys = list(state_dict_optimizer_state_keys - training_session_optimizer_state_keys)
+                if len(unexpected_keys):
+                    raise RuntimeError("Unexpected keys: {} in state_dict[optimizer][{}]"\
+                        .format(unexpected_keys, model_state_key))
+                missing_keys = list(training_session_optimizer_state_keys - state_dict_optimizer_state_keys)
+                if len(missing_keys):
+                    raise RuntimeError("Missing keys: {} in state_dict[optimizer][{}]"\
+                        .format(missing_keys, model_state_key))
+
+        def _check_key_mismatch(training_session_state_dict, state_dict):
+            model = 'model'
+            optimizer = 'optimizer'
+            # check presence of 'model' in the input state_dict
+            if model not in state_dict:
+                raise RuntimeError("Missing key: model in state_dict")
+            # check presence of 'optimizer' in the input state_dict
+            if optimizer not in state_dict:
```

Comment:
are we not allowing just loading model state or just optimizer state

Reply:
We will allow it as long as ```strict``` is false. If strict is ```True``` and the training session does exist, then we expect the complete state dictionary to be provided.

Reply:
can we associate the 'strictness' to model/optimizer state keys instead? This would be similar to pytorch where the user can load just the model state strictly

Reply:
I can do this. In case any one of the keys ```'model'``` or ```'optimizer'``` is missing. I can print a warning.